pilot/pkg/config/coredatamodel/controller.go

```diff
@@ -295,6 +261,72 @@ func (c *Controller) Delete(typ, name, namespace string) error {
 	return errUnsupported
 }
 
+func (c *Controller) serviceEntryEvents(currentStore, prevStore map[string]map[string]model.Config) {
+	dispatch := func(model model.Config, event model.Event) {}
+	if handlers, ok := c.eventHandlers[model.ServiceEntry.Type]; ok {
+		dispatch = func(model model.Config, event model.Event) {
+			log.Debugf("MCP event dispatch: key=%v event=%v", model.Key(), event.String())
+			for _, handler := range handlers {
+				handler(model, event)
+			}
+		}
+	}
+
+	// add/update
+	for namespace, byName := range currentStore {
+		for name, config := range byName {
+			if prevByNamespace, ok := prevStore[namespace]; ok {
+				if prevConfig, ok := prevByNamespace[name]; ok {
+					if config.ResourceVersion != prevConfig.ResourceVersion {
+						dispatch(config, model.EventUpdate)
+					}
+				} else {
+					dispatch(config, model.EventAdd)
+				}
+			} else {
+				dispatch(config, model.EventAdd)
+			}
+		}
+	}
+
+	// remove
+	for namespace, prevByName := range prevStore {
+		for name, prevConfig := range prevByName {
+			if byNamespace, ok := currentStore[namespace]; ok {
+				if _, ok := byNamespace[name]; !ok {
+					dispatch(prevConfig, model.EventDelete)
+				}
+			} else {
+				dispatch(prevConfig, model.EventDelete)
+			}
+		}
+	}
+}
+
+func (c *Controller) discoveryServerClearCacheEvent(descriptorType string, currentStore, prevStore map[string]map[string]model.Config) {
+	// dummy event since it is ignored by the caller
+	dispatch := func(model model.Config, event model.Event) {}
+	if handlers, ok := c.eventHandlers[descriptorType]; ok {
+		dispatch = func(model model.Config, event model.Event) {
+			log.Debugf("MCP event dispatch: key=%v event=%v", model.Key(), event.String())
+			for _, handler := range handlers {
+				handler(model, event)
+			}
+		}
+	}
+	for namespace, prevByName := range prevStore {
+		for name, prevConfig := range prevByName {
+			if byNamespace, ok := currentStore[namespace]; ok {
+				if _, ok := byNamespace[name]; !ok {
+					dispatch(prevConfig, model.EventUpdate)
```

Comment:
Costin, can you clarify what you mean by "properly implemented debounce" in the context of Pilot?

Reply:
I was under assumption that DiscoveryServer registers a clearCache event (which triggers ConfigUpdate) for all the types in configController [here](https://github.com/istio/istio/blob/565aaa99dcfc7d584e9f8eb9a04844b0b0faa1d5/pilot/pkg/proxy/envoy/v2/discovery.go#L226-L233). That's why I'm looping over those handlers above. I guess is that not the approach we want to take? do you want to call the `ConfigUpdate` by injecting the DiscoveryServer directly into coredatamodel controller? Also could you elaborate on debounce implementation in MCP server? I don't think I have come a cross that in any of our MCP server implementations.

Reply:
The handlers are some very old code - in reality model.ConfigUpdater is all that needs to be called, and only once.  It indicates that a global config recomputation is needed - which is all we support for non-EDS calls.

Calling it multiple times is wasteful. We have some complex code to deal with multiple calls - because we need to debounce config updates from K8S, but Galley is supposed to do this. 

@ayj  'debounce' means that if multiple changes happen in k8s ( like kubectl apply would produce), we don't get one event and push for each resource. 
We can still use the debounce code in pilot if it is missing in galley.

Reply:
I understand what 'debounce' means in general. I was hoping to better understand Pilot's expectations around how much to debounce. 

Galley's [publishing strategy](https://github.com/istio/istio/blob/b8e30e07ec7b36cf1ba1b445b78c81fdf0d417b9/galley/pkg/runtime/strategy.go#L22-L31) already handles debounce and is tunable.

Reply:
We use 100 ms in pilot - looks like it's similar semantic. 1s is probably ok for config changes, but for ServiceEntry endpoint updates it'll probably be far too much, we want endpoint full propagation to be <1s. 

Adding @nmittler - this needs to be taken into account for EDS-over-MCP.
The incremental EDS does not debounce EDS changes at all - they are propagated immediately.

Reply:
We've discussed creating a separate snapshot group with its own publishing strategy for endpoints.